dev/src/rate-limiter.ts

```diff
@@ -0,0 +1,161 @@
+   * @param startTime Used for testing the limiter.
+   */
+  constructor(
+    private readonly initialCapacity: number,
+    private readonly multiplier: number,
+    private readonly multiplierMillis: number,
+    private readonly startTime = Timestamp.now()
+  ) {
+    this.availableTokens = initialCapacity;
+    this.lastRefillTime = startTime;
+  }
+
+  /**
+   * Tries to make the number of operations. Returns true if the request succeeded
+   * and false otherwise.
+   *
+   * @param currentTime Used for testing the limiter.
+   * @private
+   */
+  tryMakeRequest(numOperations: number) {
+    return this._tryMakeRequest(numOperations, Timestamp.now());
+  }
+
+  /**
+   * Used in testing for different timestamps.
+   *
+   * @private
+   */
+  // Visible for testing.
+  _tryMakeRequest(numOperations: number, currentTime: Timestamp): boolean {
+    this.refillTokens(currentTime);
+    if (numOperations <= this.availableTokens) {
+      this.availableTokens -= numOperations;
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Returns the number of ms needed to make a request with the provided number
+   * of operations. Returns 0 if the request can be made with the existing
+   * capacity. Returns -1 if the request is not possible with the current
+   * capacity.
+   *
+   * @private
+   */
+  getNextRequestDelayMs(numOperations: number) {
+    return this._getNextRequestDelayMs(numOperations, Timestamp.now());
+  }
+
+  /**
+   * Used in testing for different timestamps.
+   *
+   * @private
+   */
+  // Visible for testing.
+  _getNextRequestDelayMs(
+    numOperations: number,
+    currentTime: Timestamp
+  ): number {
+    if (numOperations < this.availableTokens) {
+      return 0;
+    }
+
+    const capacity = this.calculateCapacity(currentTime, this.startTime);
+    if (capacity < numOperations) {
+      return -1;
+    }
+
+    const requiredTokens = numOperations - this.availableTokens;
+    return Math.ceil((requiredTokens * 1000) / capacity);
+  }
+
+  /**
+   * Refills the number of available tokens based on how much time has elapsed
+   * since the last time the tokens were refilled.
+   *
+   * @private
+   */
+  private refillTokens(currentTime = Timestamp.now()): void {
+    if (currentTime.toMillis() > this.lastRefillTime.toMillis()) {
+      const elapsedTime =
+        currentTime.toMillis() - this.lastRefillTime.toMillis();
+      const capacity = this.calculateCapacity(currentTime, this.startTime);
+      const tokensToAdd = Math.floor((elapsedTime * capacity) / 1000);
+      if (tokensToAdd > 0) {
+        this.availableTokens = Math.min(
+          capacity,
+          this.availableTokens + tokensToAdd
+        );
+        this.lastRefillTime = currentTime;
+      }
+    }
+  }
+
+  /**
+   * Calculates the maximum capacity based on the two provided timestamps.
+   *
+   * @private
+   */
+  // Visible for testing.
+  calculateCapacity(currentTime: Timestamp, startTime: Timestamp): number {
```

Comment:
Is it possible to drop the `startTime` argument here? It makes this code easier to reason about. In the tests, you would simply have to create a new instance.